subprojects/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/DefaultArtifactSelector.java

```diff
@@ -99,29 +102,31 @@ private ResolvedVariant toResolvedVariant(VariantResolveMetadata.Identifier iden
                                               ExcludeSpec exclusions,
                                               ModuleVersionIdentifier ownerId,
                                               ModuleSources moduleSources,
-                                              @Nullable Map<VariantResolveMetadata.Identifier, ResolvedVariant> resolvedVariantCache) {
+                                              ResolvedVariantCache resolvedVariantCache) {
         // artifactsToResolve are those not excluded by their owning module
         List<? extends ComponentArtifactMetadata> artifactsToResolve = CollectionUtils.filter(artifacts,
                 artifact -> !exclusions.excludesArtifact(ownerId.getModule(), artifact.getName())
         );
 
         boolean hasExcludedArtifact = artifactsToResolve.size() < artifacts.size();
+        ImmutableAttributes attributes = artifactTypeRegistry.mapAttributesFor(variantAttributes, artifacts);
 
         if (hasExcludedArtifact || identifier == null) {
             // An ad hoc variant, has no identifier
-            return createResolvedVariant(null, displayName, variantAttributes, artifacts, capabilities, ownerId, moduleSources, artifactsToResolve);
+            return createResolvedVariant(null, displayName, attributes, capabilities, ownerId, moduleSources, artifactsToResolve);
         } else {
-            if (resolvedVariantCache == null) {
-                return createResolvedVariant(identifier, displayName, variantAttributes, artifacts, capabilities, ownerId, moduleSources, artifactsToResolve);
-            } else {
-                return resolvedVariantCache.computeIfAbsent(identifier, id -> createResolvedVariant(identifier, displayName, variantAttributes, artifacts, capabilities, ownerId, moduleSources, artifactsToResolve));
-            }
+            // This is a bit of a hack because we allow the artifactType registry to be different in every resolution scope.
+            // This means it's not safe to assume a variant resolved in one consumer can be reused in another consumer with the same key.
```

Comment:
Can we use a compound key which includes the artifact type registry or perhaps the return attributes from `artifactTypeRegistry.mapAttributesFor`?

Reply:
That's basically what `VariantWithOverloadAttributes` is. It's a key that's the identifier + any extra attributes from the artifact type.

Reply:
Ahh I see, I was confused with what `VariantWithOverloadAttributes` was actually doing, but it makes sense now given that it is used as the cache key.